#18758 Send appropriate events on remote actor system shutdown and quarantine

jdevelop opened the pull request:
Publish appropriate events to the current ActorSystem event stream upon remote ActorSystem shutdown or when current ActorSystem is quarantined by the remote ActorSystem.

jdevelop committed:
Add named exception to detect when a cluster node has been quarantined by others #18758

Publish appropriate events to the current ActorSystem event stream upon remote ActorSystem shutdown or when current ActorSystem is quarantined by the remote ActorSystem.

johanandren commented:
OK TO TEST

johanandren commented:
Refs #18758

johanandren commented:
Should have test coverage?

johanandren commented:
And probably documentation as well.

johanandren commented:
PLS BUILD

jdevelop commented:
Anyone has an idea of how can I emulate that disassociation event in a sort of integration test? Or setting up an instance of Endpoint and sending a disassociation message to it would be enough? Please advice.

johanandren commented:
Maybe ```NodeLeavingAndExitingAndBeingRemovedSpec``` in akka-cluster can be useful to look at?

johanandren commented:
PLS BUILD

jdevelop commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/RemoteRestartedQuarantinedSpec.scala:
```diff
@@ -114,6 +115,10 @@ abstract class RemoteRestartedQuarantinedSpec
 
+        expectMsgPF(10 seconds) {
+          case AssociationHandle.ActorSystemQuarantinedEvent(local, remote) ⇒
+        }
```
I have added this expectation just to test that the message is received upon quarantine. Not sure if any further validation is necessary.

jdevelop commented:
@johanandren I realized that there is the multi-jvm test for the quarantined event, so I altered it to watch for the new event introduced.

johanandren commented:
Just realized the ticket was [triaged] but there was little discussion about how this should be achieved, so I'll stop feedbacking it for now until someone from the Akka-team says that this way to solve the ticket is a good idea, so that you do not have to spend a lot of time fixing it up in vain.

jdevelop commented:
Okay, I'll add some description to the ticket and will see. Thanks for the help!

rkuhn commented on akka-remote/src/main/scala/akka/remote/transport/Transport.scala:
```diff
@@ -179,6 +179,9 @@ object AssociationHandle {
   case object Quarantined extends DisassociateInfo
 
+  case class ActorSystemQuarantinedEvent(localAddress: Address, remoteAddress: Address)
+  case class RemoteShutdownEvent(localAddress: Address, remoteAddress: Address)
```
These are user API and should live directly in `akka.remote`, plus it would be helpful to add ScalaDoc to them.

rkuhn commented:
Thanks for the initiative! I think this is the right way to go about this, please add docs [here](http://doc.akka.io/docs/akka/2.4.1/scala/remoting.html#Remote_Events) and we can merge it.

jdevelop commented:
@rkuhn ok, thanks for the hint! Please take a look at the updates in my PR.

rkuhn commented on akka-remote/src/main/scala/akka/remote/RemotingLifecycleEvent.scala:
```diff
@@ -87,6 +87,9 @@ final case class QuarantinedEvent(address: Address, uid: Int) extends RemotingLi
 }
 
+@SerialVersionUID(1L)
+final case class ActorSystemQuarantinedEvent(localAddress: Address, remoteAddress: Address)
```
Please also extend RemotingLifecycleEvent and add a descriptive `toString` message. In order to distinguish the meaning from QuarantinedEvent this one should be called something like WeHaveBeenQuarantinedEvent (my naming skills are not perfect today, so if you have a better idea … ). I think it should be logged at warning level and replace the old unstructured log message.

jdevelop replied:
@rkuhn Done, pls take a look

rkuhn commented:
Looking good, apart from that now there should be two warnings logged when quarantined, right?

jdevelop commented:
@rkuhn I'm not sure about the logging. If you mean the following code

    case AssociationHandle.Quarantined ⇒
      context.system.eventStream.publish(ThisActorSystemQuarantinedEvent(localAddress, remoteAddress))
      throw InvalidAssociation(
        localAddress,
        remoteAddress,
        InvalidAssociationException("The remote system has quarantined this system. No further associations " +
          "to the remote system are possible until this system is restarted."))

it seems that this exception would cause the actor to terminate, because it's not handled anywhere. So perhaps this one is required for the correct workflow.

rkuhn commented:
Hmm, trying to figure out what is logged where (and how that other log message gets printed that appears in the test case above your addition) I am becoming less sure of this approach: should this new event not be published from the EndpointManager’s supervisorStrategy instead? /cc @drewhk

jdevelop commented:
@rkuhn the new event is not supposed to be a drop-in replacement for the current workflow of supervisor actor. The only goal is to get the notification that the current system is quarantined, to the user may take some countermeasurements (restart system, restart app etc). That's why I didn't remove the current "throw an error and let the supervision strategy to deal with it) but added some notification to the event stream.

rkuhn commented:
Right, but the supervisorStrategy also publishes to the eventStream, so it might be more appropriate to hook it in there. I’ll wait until Endre chimes in.

drewhk commented:
I would also like to hook into the supervision in the Remoting class instead (since that is where these exceptions bubble up) and put the notification there.

jdevelop commented:
@drewhk , @rkuhn ok, so just to confirm - the event should be propagated not in the Endpoint, but in Remoting. Anything else?

rkuhn commented:
Yes, that is the only change, thanks!

jdevelop commented:
@rkuhn , @drewhk 
I moved the event notification to the Remoting supervisor, also I added the disassociation info to  InvalidAssociation class, so it will be possible to see what exactly went wrong (instead of picking up a string message). Please review.

rkuhn commented:
OK TO TEST

jdevelop commented:
@rkuhn  I wonder if this failure means something about my changes:

    java.lang.RuntimeException: akka-remote: Binary compatibility check failed!
	at scala.sys.package$.error(package.scala:27)
	at com.typesafe.tools.mima.plugin.SbtMima$$anonfun$reportErrors$1.apply(SbtMima.scala:77)
 	at com.typesafe.tools.mima.plugin.SbtMima$$anonfun$reportErrors$1.apply(SbtMima.scala:62)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:224)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:403)
	at com.typesafe.tools.mima.plugin.SbtMima$.reportErrors(SbtMima.scala:62)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$4.apply(MimaPlugin.scala:32)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$4.apply(MimaPlugin.scala:31)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)

Is it something I have to fix?

drewhk commented:
It is a false positive about binary compatibility. A problem filter needs to be added to the build.

jdevelop commented:
@drewhk still not clear - should I do something about this?

rkuhn commented:
Thanks @jdevelop !

jdevelop commented:
Perhaps somebody could close the issue #18758
